qiskit_nature/operators/second_quantization/fermionic_op.py

```diff
@@ -464,6 +465,74 @@ def to_list(
             ]
         return self._to_dense_label_data()
 
+    def to_matrix(self, sparse: Optional[bool] = True) -> Union[csc_matrix, np.ndarray]:
+        """Convert to a matrix representation over the full fermionic Fock space in occupation number
+        basis. The basis states are ordered in increasing bitstring order as 0000, 0001, ..., 1111.
+
+        Args:
+            sparse: If true, the matrix is returned as a sparse csc_matrix, else it is returned as a
+            dense numpy array.
+
+        Returns:
+            The matrix of the operator in the Fock basis (scipy.sparse.csc_matrix or numpy.ndarray
+            with dtype=numpy.complex128)
+        """
+
+        csc_data, csc_col, csc_row = [], [], []
+
+        # loop over all columns of the matrix
+        for col_idx in range(2 ** self.register_length):
+            initial_occupations = [
+                int(occ) for occ in np.binary_repr(col_idx, self.register_length)
+            ]
+            # loop over the terms in the operator data
+            for opstring, prefactor in self.reduce()._data:
+                # check if op string is the identity
+                if not opstring:
+                    csc_data.append(prefactor)
+                    csc_row.append(col_idx)
+                    csc_col.append(col_idx)
+                else:
+                    occupations = initial_occupations.copy()
+                    sign = 1
+                    mapped_to_zero = False
+
+                    # apply terms sequentially to the current basis state
+                    for label_primitive in reversed(opstring):
+                        if not label_primitive.is_creation:
+                            # If this mode is not occupied, the action of '-' on the state is zero
+                            if occupations[label_primitive.index] == 0:
+                                mapped_to_zero = True
+                                break
+                            sign *= (-1) ** sum(occupations[: label_primitive.index])
+                            occupations[label_primitive.index] = 0
+
+                        if label_primitive.is_creation:
+                            # If this mode is already occupied, the action of '+' on this state is zero
+                            if occupations[label_primitive.index] == 1:
+                                mapped_to_zero = True
+                                break
+                            sign *= (-1) ** sum(occupations[: label_primitive.index])
+                            occupations[label_primitive.index] = 1
```

Comment:
Assuming you change `occupations` to store booleans rather than integers you can simplify the code
by quite a bit until it becomes the below:
```suggestion
                        occ = occupations[label_primitive.index]
                        if label_primitive.is_creation == occ:
                            # Applying the creation operator on an occupied state maps to zero. So
                            # does applying the annihilation operator on an unoccupied state.
                            mapped_to_zero = True
                            break
                        sign *= (-1) ** sum(occupations[: label_primitive.index])
                        occupations[label_primitive.index] = not occ
```